src/portfft/committed_descriptor_impl.hpp

```diff
@@ -534,72 +638,53 @@ class committed_descriptor_impl {
-        }
-      } else {
-        if (compute_direction == direction::BACKWARD) {
-          conjugate_on_load = detail::complex_conjugate::APPLIED;
-          conjugate_on_store = detail::complex_conjugate::APPLIED;
-        }
-        set_spec_constants(level, in_bundle, params.lengths[dimension_num], factors,
+    std::vector<input_bundles_and_metadata_t> input_kernels_and_metadata;
+    for (const auto& [level, kernel_ids, factors] : prepared_vec) {
+      input_kernels_and_metadata.emplace_back(
+          level, sycl::get_kernel_bundle<sycl::bundle_state::input>(queue.get_context(), kernel_ids), factors);
+    }
+    bool is_compatible = true;
+    if (top_level == detail::level::GLOBAL) {
+      set_global_impl_spec_constants(input_kernels_and_metadata, num_forward_factors, num_backward_factors,
+                                     compute_direction, num_backward_factors > 0, scale_factor);
+    } else {
+      detail::complex_conjugate conjugate_on_load = detail::complex_conjugate::NOT_APPLIED;
+      detail::complex_conjugate conjugate_on_store = detail::complex_conjugate::NOT_APPLIED;
+      detail::apply_scale_factor scale_factor_applied = detail::apply_scale_factor::APPLIED;
+      if (compute_direction == direction::BACKWARD) {
+        conjugate_on_load = detail::complex_conjugate::APPLIED;
+        conjugate_on_store = detail::complex_conjugate::APPLIED;
+      }
+      if (skip_scaling) {
+        scale_factor_applied = detail::apply_scale_factor::NOT_APPLIED;
+      }
+      for (auto& [level, input_bundle, factors] : input_kernels_and_metadata) {
+        set_spec_constants(level, input_bundle, params.lengths[dimension_num], factors,
                            detail::elementwise_multiply::NOT_APPLIED, detail::elementwise_multiply::NOT_APPLIED,
                            scale_factor_applied, level, conjugate_on_load, conjugate_on_store, scale_factor);
       }
+    }
+
+    for (const auto& [level, input_bundle, factors] : input_kernels_and_metadata) {
       try {
-        PORTFFT_LOG_TRACE("Building kernel bundle with subgroup size", SubgroupSize);
-        result.emplace_back(sycl::build(in_bundle), factors, params.lengths[dimension_num], SubgroupSize,
-                            PORTFFT_SGS_IN_WG, std::shared_ptr<Scalar>(), level);
-        PORTFFT_LOG_TRACE("Kernel bundle build complete.");
-      } catch (std::exception& e) {
+        result.emplace_back(
+            sycl::build(input_bundle), factors,
+            static_cast<std::size_t>(std::accumulate(factors.begin(), factors.end(), 1, std::multiplies<Idx>())),
+            SubgroupSize, PORTFFT_SGS_IN_WG, std::shared_ptr<Scalar>(), level);
+      } catch (const std::exception& e) {
         PORTFFT_LOG_WARNING("Build for subgroup size", SubgroupSize, "failed with message:\n", e.what());
         is_compatible = false;
-        break;
       }
-      counter++;
-    }
-    if (is_compatible) {
-      return result;
+      if (!is_compatible) {
```

Comment:
`is_compatible` isn't needed, just return nullopt immediately where it is set to false